File: gymgenie-mobile/composeApp/src/androidMain/kotlin/com/asc/gymgenie/utils/DayUtils.kt
package com.asc.gymgenie.utils

import java.util.Calendar

/**
 * Centralized weekday name translations and ordering constants.
 * All backend wire values are uppercase English (e.g. "MONDAY").
 */

/** Canonical backend ordering. */
val WeekdayOrder: List<String> = listOf(
    "MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY", "SATURDAY", "SUNDAY",
)

/** Ordered pairs of (backend wire, short Russian label). */
val WeekdayPairs: List<Pair<String, String>> = listOf(
    "MONDAY" to "Пн",
    "TUESDAY" to "Вт",
    "WEDNESDAY" to "Ср",
    "THURSDAY" to "Чт",
    "FRIDAY" to "Пт",
    "SATURDAY" to "Сб",
    "SUNDAY" to "Вс",
)

/** Ordered short Russian day labels: ["Пн", "Вт", ...] */
val WeekdayLabelsRu: List<String> = WeekdayPairs.map { it.second }

/** Full Russian day name from backend wire value (e.g. "MONDAY" -> "понедельник"). */
fun weekdayNameRu(wire: String): String = when (wire.uppercase()) {
    "MONDAY" -> "понедельник"
    "TUESDAY" -> "вторник"
    "WEDNESDAY" -> "среда"
    "THURSDAY" -> "четверг"
    "FRIDAY" -> "пятница"
    "SATURDAY" -> "суббота"
    "SUNDAY" -> "воскресенье"
    else -> wire.lowercase()
}

/** Short Russian day abbreviation from backend wire value (e.g. "MONDAY" -> "Пн"). */
fun weekdayShortRu(wire: String): String = when (wire.uppercase()) {
    "MONDAY" -> "Пн"
    "TUESDAY" -> "Вт"
    "WEDNESDAY" -> "Ср"
    "THURSDAY" -> "Чт"
    "FRIDAY" -> "Пт"
    "SATURDAY" -> "Сб"
    "SUNDAY" -> "Вс"
    else -> wire
}

/** Short Russian day abbreviation from [Calendar] day constant (e.g. [Calendar.MONDAY] -> "Пн"). */
fun weekdayShortFromCalendar(calendarDay: Int): String = when (calendarDay) {
    Calendar.MONDAY -> "Пн"
    Calendar.TUESDAY -> "Вт"
    Calendar.WEDNESDAY -> "Ср"
    Calendar.THURSDAY -> "Чт"
    Calendar.FRIDAY -> "Пт"
    Calendar.SATURDAY -> "Сб"
    Calendar.SUNDAY -> "Вс"
    else -> ""
}

/** Short Russian day abbreviation from [java.time.DayOfWeek]. */
fun weekdayShortFromDayOfWeek(day: java.time.DayOfWeek): String = when (day) {
    java.time.DayOfWeek.MONDAY -> "Пн"
    java.time.DayOfWeek.TUESDAY -> "Вт"
    java.time.DayOfWeek.WEDNESDAY -> "Ср"
    java.time.DayOfWeek.THURSDAY -> "Чт"
    java.time.DayOfWeek.FRIDAY -> "Пт"
    java.time.DayOfWeek.SATURDAY -> "Сб"
    java.time.DayOfWeek.SUNDAY -> "Вс"
}

/** Short Russian day abbreviation from [kotlinx.datetime.DayOfWeek]. */
fun weekdayShortFromKotlinxDayOfWeek(day: kotlinx.datetime.DayOfWeek): String = when (day) {
    kotlinx.datetime.DayOfWeek.MONDAY -> "Пн"
    kotlinx.datetime.DayOfWeek.TUESDAY -> "Вт"
    kotlinx.datetime.DayOfWeek.WEDNESDAY -> "Ср"
    kotlinx.datetime.DayOfWeek.THURSDAY -> "Чт"
    kotlinx.datetime.DayOfWeek.FRIDAY -> "Пт"
    kotlinx.datetime.DayOfWeek.SATURDAY -> "Сб"
    kotlinx.datetime.DayOfWeek.SUNDAY -> "Вс"
}

/** Reverse mapping: short Russian label to backend wire value (e.g. "Пн" -> "MONDAY"). */
val dayLabelToBackend: LinkedHashMap<String, String> = linkedMapOf(
    "Пн" to "MONDAY",
    "Вт" to "TUESDAY",
    "Ср" to "WEDNESDAY",
    "Чт" to "THURSDAY",
    "Пт" to "FRIDAY",
    "Сб" to "SATURDAY",
    "Вс" to "SUNDAY",
)

/** Reverse mapping: backend wire value to short Russian label (e.g. "MONDAY" -> "Пн"). */
val backendToDayLabel: Map<String, String> = dayLabelToBackend.entries.associate { (k, v) -> v to k }

File: gymgenie-mobile/iosApp/iosApp/Utils/DayUtils.swift
import Foundation

// MARK: - Backend day name → full Russian name

/// Converts a backend weekday constant (e.g. "MONDAY") to its full Russian name.
func dayNameRu(_ day: String) -> String {
    switch day.uppercased() {
    case "MONDAY": return "понедельник"
    case "TUESDAY": return "вторник"
    case "WEDNESDAY": return "среда"
    case "THURSDAY": return "четверг"
    case "FRIDAY": return "пятница"
    case "SATURDAY": return "суббота"
    case "SUNDAY": return "воскресенье"
    default: return day.lowercased()
    }
}

// MARK: - Backend day name → short Russian abbreviation

/// Converts a backend weekday constant (e.g. "MONDAY") to a short Russian label ("Пн").
func dayShortRu(_ day: String) -> String {
    switch day.uppercased() {
    case "MONDAY": return "Пн"
    case "TUESDAY": return "Вт"
    case "WEDNESDAY": return "Ср"
    case "THURSDAY": return "Чт"
    case "FRIDAY": return "Пт"
    case "SATURDAY": return "Сб"
    case "SUNDAY": return "Вс"
    default: return day
    }
}

// MARK: - Calendar weekday index → short Russian abbreviation

/// Converts a `Calendar.component(.weekday)` value (1 = Sunday … 7 = Saturday) to "Пн", "Вт", etc.
func dayShortRuFromCalendarIndex(_ index: Int) -> String {
    switch index {
    case 1: return "Вс"
    case 2: return "Пн"
    case 3: return "Вт"
    case 4: return "Ср"
    case 5: return "Чт"
    case 6: return "Пт"
    case 7: return "Сб"
    default: return ""
    }
}

// MARK: - Calendar weekday index → backend day constant

/// Converts `Calendar.component(.weekday)` (1 = Sunday) to the backend enum string "MONDAY", etc.
func calendarWeekdayToBackend(_ weekday: Int) -> String {
    switch weekday {
    case 1: return "SUNDAY"
    case 2: return "MONDAY"
    case 3: return "TUESDAY"
    case 4: return "WEDNESDAY"
    case 5: return "THURSDAY"
    case 6: return "FRIDAY"
    case 7: return "SATURDAY"
    default: return "MONDAY"
    }
}

// MARK: - Lookup tables

/// Maps backend day constants to short Russian labels.
let backendDayToShort: [String: String] = [
    "MONDAY": "Пн", "TUESDAY": "Вт", "WEDNESDAY": "Ср",
    "THURSDAY": "Чт", "FRIDAY": "Пт", "SATURDAY": "Сб", "SUNDAY": "Вс"
]

/// Maps short Russian labels to backend day constants.
let dayLabelToBackend: [String: String] = [
    "Пн": "MONDAY", "Вт": "TUESDAY", "Ср": "WEDNESDAY",
    "Чт": "THURSDAY", "Пт": "FRIDAY", "Сб": "SATURDAY", "Вс": "SUNDAY"
]

/// Canonical Monday-first backend day ordering.
let weekdayOrder: [String] = [
    "MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY", "SATURDAY", "SUNDAY"
]

/// Short Russian day labels in Monday-first order, matching `weekdayOrder`.
let daysOfWeekLabels: [String] = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"]
